src/Peachpie.CodeAnalysis/FlowAnalysis/ExpressionAnalysis.cs

```diff
@@ -1416,6 +1498,19 @@ void BindParam(PhpParam expected, BoundArgument givenarg)
                     // TODO: Err, variable or field must be passed into byref argument. foo("hello") where function foo(&$x){}
                 }
             }
+            else if (!expected.Type.IsAnyType && givenarg.Value is BoundVariableRef refvar && refvar.Name.IsDirect)
+            {
+                // Even for variables passed by value we may gain information about their type (if we previously had none),
+                // because not complying with the parameter type would have caused throwing a TypeError
+                var local = State.GetLocalHandle(refvar.Name.NameValue);
+                var localType = State.GetLocalType(local);
+                if (localType.IsAnyType && !localType.IsRef &&
```

Comment:
why `!localType.IsRef` ?

Reply:
```php
function foo(array $a, &$b) {
  $b = "foo";
}

function test() {
  $a = [42];
  $b =& $a;
  foo($a, $b);
  echo $a;
}

test();
```
`$a` then contains the string instead of the original array.